README.md
# HTML Links Checker Extension for Visual Studio Code

This is a Visual Studio Code extension for HTML coding.

## Overview
This extension validates URLs inside HTML code (A tags) of the Active Editor.

For the moment, this extension checks:
+ if anchor link is present in the current code;
+ if root relative link is working (link that starts with ```/``` like ```/folder/file.html```)
+ if link is redirected and what's the new URL (redirection)
+ if link is NOT working (including links that require authentication)
+ if A tag is empty (no href or empty href)
+ if A tag has no text (ex.: ```<a href="abc.com"></a>```)
+ if accessibility requirements are set for external link.

The extension can't validate links that:
+ require authentication
+ are emails
+ are relative from the folder location (ex.: ```../folder/file.html``` OR ```file.html```)
+ are FTP
+ point to local file (```file://```)

## How to use
The extension uses the selected text or the whole content (if no selection).

Trigger the extension by using <kbd>CTRL</kbd>+<kbd>SHIFT</kbd>+<kbd>P</kbd> and typing "html link checker".
Use the Visual Studio Code's **Problems tab** to review errors/warnings and apply appropriate action.

When an error/warning is fixed, the extension must be re-run to remove the link from the Probloms tab. 


## Configuration instructions
Following are options with descriptions and instructions

### Request Method
Select the method used to send the request to the URL.

Available options are: **"head"**, "get" or "post.


### Local Domain (mandatory)
Single string that represent the local domain name server. It's added before root relative links. Must include the protocol and no slash at the end (ex.: ```http://www.mydomain.com```, ```https://mydomain.com```)


### Exluded Domains (optional)
If used, it must be defined in JSON format. It's an array of domain names to not validate. Only include the domaine name without any protocol (ex.: domainname.com)
```JSON
"html-links-checker.excludedDomains": [
  "google.com",
  "yourdomainname.com"
]
```


### Validate External Link Accessibility
Boolean value to indicate if the accessibility for external link should processed.
Default value: True


### External Link Text
Array of string.
Link accessibility suggests to add term in link's name to precise that the link is an external link. The users can know that they will quit the site to go somewhere else.
Usualy, in english we add "external link" (hidden or not)
Default value: 
  - external link
  - lien externe

To add/remove/modify list you must set it in settings directly in JSON
```JSON
"html-links-checker.externalLinkText": {
  "description": "Specifies string that should be included in external link text to be accessible",
  "type": "array",
  "default": [
    "external link",
    "lien externe"
  ]
}
```

### Show Protocol Redirection Warning
Define how to manage redirection caused only by protocol change (ex.: ```http://``` to ```https://```).
Available Options:
  + "yes - separate" : A warning will be added no matter if another error occur for the same link
  + "Yes - global" : A warning will be added only if no other error are related to this link
  + "No" : No warning will be added for this kind of redirection.


### Show WWW Redirection Warning
Define how to mange redirection caused only by WWW change (if WWW has been added or removed from the initial URL).
  + "yes - separate" : A warning will be added no matter if another error occur for the same link
  + "Yes - global" : A warning will be added only if no other error are related to this link
  + "No" : No warning will be added for this kind of redirection.


## TODO
+ Validate relative link from folder:
  + Adding a Extension property to define the full local path to site root folder
  + Create the function that validate if file exists localy (no web requets)
+ internal link:
  + adding multiple local domains instead of only one. Allow to validate the DEV and PROD servers, and indicate if the links is only working on the DEV server.
  + if both local servers not return a good result (status 200), then check if the file exists localy in file system.
  + if not get a status 200 on PROD server, all other ways used to valide the link should be shown to user to take appropriate action on link.
+ validate URL in images
+ check if possible to quickly replace redirected link by the new link
+ adding popups to get username and password for link that requires authentication. Maybe adding Extension's properties to store username and password (insecur).
+ find a way to validate email (if possible)
+ validate FTP link


## License
[MIT](https://choosealicense.com/licenses/mit/)

CHANGELOG.md
# Change Log

## [1.1.14] - 2021-XXXXXXXXXXXXXXXXXXXXXXXXXXX


## [1.1.13] - 2021-04-14
+ Fix issue with External link validation (no validation proceeded)


## [1.1.12] - 2021-02-15
+ Add external link accessibility validation with 2 new Extensions's proterties to enabled/disable this validation and define the text to check<br>
  Usually an external link, to be accessible, must have an attribute ```rel="external"``` and an hidden text in the name's link to say that the link will bring the user outside the current web site, the common term used is: "external link"
+ Fix an issue with local file (file://) and ftp links that bug the progression.

## [1.1.11] - 2021-02-10
+ Downgrade the available version of VSCode in package.json for more compatibility

## [1.1.10] - 2021-02-09
+ Converted urllib with the Promise way
+ Added delay to each 8 links to not attach the DNS <br>
  1 second to each 8 links processes<br>
  To be improved but work better than before<br>
  2 new Extension's properties added to manage the number of links and the delay (milisecond)
+ Added a "completed" message at the end of process

## [1.1.9] - 2021-01-27
+ Tried to fix an issue with response with error and status code 200. It seems that status code is 200 but timeout occur and generates an error. So, the URL is 200 (valid) but is in error due to timeout.
+ Test for proxy connexion.

## [1.1.8] - 2021-01-27
+ Fixed issue related to "this" inside a "forEach"
+ Update comments in code

## [1.1.2 to 1.1.7] - 2021-01-27
+ Tests to fix issue on Windows

## [1.1.1] - 2021-01-26
+ Updated dependencies

## [1.1.0] - 2021-01-25
**First attempt with this new rebuilt code.**<br>
Working version to test/evaluate in production to find improvement before to continue code conversion.
+ Rebuilt the code with object perspective
+ Multiple Error/Warning messages in 'Problems' tab are merged together to avoid duplication of message for the same link. Only one message with all errors/warnings in it.
+ Warning added for relative URL, because this kind of link is not validate for the moment (need manual validation)
+ Email validation (format) has been added with error message if the address is not a valide email format.
+ Validation of spaces in URL, error added when spaces is present in URL because it's not supposed to have space in URL, %20 must be replace spaces.

## [1.0.8] - 2021-01-12
+ Test to fix working link but displayed as error.<br>
  It's temporary. The code is reviewed and big change will come.

## [1.0.7] - 2021-01-08
+ Remove validation of space in local file (FILE:)
+ changes some useless part of code

## [1.0.6] - 2021-01-08
+ Trying to fix a bug on Windows that don't have on Linux

## [1.0.5] - 2021-01-08
+ Little bug fixed and adding a temporary variable watcher 

## [1.0.4] - 2021-01-07
+ Extension's property name fixed in code.

## [1.0.3] - 2021-01-06
+ Change node module for URL request from "needle" to "urllib".<br />More powerfull and quicker for my needs
+ Added Extensions's properties to have a better way to display link warning/error
  + Protocol: This setting is related to redirection caused only by changing the protocl. Example: from ```http``` to ```https```<br />allows to force warning in any case, only display if it's the only error, or no display warning.
  + WWW: This setting is related to redirection caused only by adding ```www``` at the domain name.<br />allows to force warning in any case, only display if it's the only error, or no display warning.
  + Request method: Allow to set the prefered method to use to send request: "head", "get" or "post"
+ Only use the prefered request method instead of re-request URL with "get" when not working with "head". Maybe will re-add this feature later. Future will say.



## [1.0.2] - 2021-01-04
+ Allow validation in a part of the code (selected text).
  If no text selected the whole content is selected.
+ Minors fixes and improvement in the code
+ Add a Information in the 'Problems' tab when link is part of the excluded domains list.


## [1.0.1] - 2020-12-23
Fixed issue with relative links.
The bug was with URL without protocol (http://) and short domain (ex.: "domain.com"). It was considered as a relative URL instead of URL without protocol.
The regex realted to "relative" link has been updated to include file type (html, htm, asp, php, txt).

Added validation of A tag without text (ex.: <a href="domain.name"></a>).
An error is generated if it's occur.


## [1.0.0] - 2020-12-23
Initial version of extension is able to validate:
+ anchor
+ internal link relative from root: start with dash (/) not with ../
+ external link

Returns an error for:
+ link that requires authentication
+ link relative from current root ('../folder/file.html' or 'file.html')
+ link not working (no domain name or status 404)
+ link redirected (provide new URL)
+ Empty A tag (no href or href is empty)
+ (optional) Accessibility for External link.

No validate:
+ email address
+ FTP link
+ local file link (FILE://)

## previous [1.0.0]
- Was for development and testing. No longer version before 1.0.0 should be used.


src/extension.js
const vscode = require('vscode');
const genFunc = require('./genericFunctions');
const urllib = require('urllib'); // https://www.npmjs.com/package/urllib
const JSDOM = require('jsdom').JSDOM;


// Create a new Collection for Diagnostics
// A Diagnostics Collection must be outsides of a function.
const links_checker_diagColl = vscode.languages.createDiagnosticCollection("html-link-checker");
// Clear all diagnostics from the "Problems" Tab when we close the file
vscode.workspace.onDidCloseTextDocument(function (listener) {
	// clear all previous diagnostics of the closed editor
	links_checker_diagColl.delete(listener.uri);
	links_checker_diagColl.set(listener.uri, undefined);
});


/** ***********************************************************************************************
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {

	let disposable = vscode.commands.registerCommand('html-links-checker.start', function () {
		// If code is not HTML, then exits with a message, message sent inside the isHTMLcode function.
		if (!genFunc.isHTMLcode()) {
			return;
		}

		// Clear previous Diagnostics
		links_checker_diagColl.set(vscode.window.activeTextEditor.document.uri, []);

		// Main process - Validate all URL 
		main_Validation_Process();
	});


	// add main function to context array
	context.subscriptions.push(disposable);
}
exports.activate = activate;

function deactivate() { }

module.exports = {
	activate,
	deactivate
}

/**
 * This is the main funcion that start the validation process
 */
async function main_Validation_Process() {
	// Create a DOM with ALL content (used to valide IDs)
	const tt = genFunc.getWholeText();
	const wholeDOM = new JSDOM(tt, { includeNodeLocations: true, contentType: "text/html" });

	// Get selected text
	const curContent = genFunc.getTextSelected(true);
	// Create a DOM from the selected text
	let myDOM = new JSDOM(curContent, { includeNodeLocations: true, contentType: "text/html" });
	// Retrieve all A tag in the DOM from selected text
	const ATags = myDOM.window.document.querySelectorAll('a');

	// *** Process each A Tag - start ****
	// Array that contains all object URLValidator
	let arrayOfProcesses = [];
	// Manager the number of link validate before to wait.
	let cpt = 0;
	// get Extension's property values for the process
	const nbIteration = vscode.workspace.getConfiguration("html-links-checker").delayNumberOfLinksBeforeWait;
	const nbMilisecond = vscode.workspace.getConfiguration("html-links-checker").delayNumberOfLinksBeforeWait;
	// Loop each A tag
	for (let x = 0; x < ATags.length; x++) {
		if (cpt == nbIteration) {
			await genFunc.waiter(nbMilisecond)
			cpt = 0;
		}
		// Create a new object in the array (object automatically validates the link)
		arrayOfProcesses.push(new URLValidator(ATags[x], wholeDOM));
		cpt++;
	}
	// *** Process each A Tag - end ****

	// This section waits until all links will have been processed
	// object property "completed" is "true" for all links.
	let flagWhile = true;
	while (flagWhile) {
		await genFunc.waiter(800)
		let flagFor = true;
		for (let x = 0; x < ATags.length; x++) {
			if (arrayOfProcesses[x].completed == false) {
				flagFor = false
			}
		}
		if (flagFor) {
			flagWhile = false;
		}
	}
	vscode.window.showInformationMessage("Processus terminé");

}


/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
/* #*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*
*/


/**
 * Object that validate an A tag element
 * TODO: change contructor to indicate if the object is an A tag or a IMG tag
 */
class URLValidator {
	/**
	 * 
	 * @param {JSDOM.nodeElem} JSDOM_Link_tag is the tag that contains the URL to check (A or IMG tag)
	 * @param {JSDOM} wholeDOM is the DOM of the whole current active editor.
	 * 												 Must be the whole DOM for the position and for the Anchor validation
	 */
	constructor(JSDOM_Link_tag, wholeDOM) {
		this.wholeDOM = wholeDOM;
		this.domTag = JSDOM_Link_tag;
		this.url = this.domTag.getAttribute('href');
		this.text = this.domTag.text;

		// URL Types properties
		this.urlType = {};

		// Error Type properties
		this.error = {};

		// Warning Type properties
		this.warning = {};

		// Variable of control
		let canProcessValidation = true;

		// When URL has been processed, no matter the result, this attribute is set to TRUE
		this.completed = false;

		// Evaluate the URL to determine its kind/preliminary error
		this.processURL();

		// If URL is empty or it's in the excluded list, no validation required
		if (this.error.emptyURL || this.urlType.isExcluded) {
			canProcessValidation = false;
		}
		// If URL contains spaces then not process it
		else if (this.url.match(/\s/gi) != null) {
			this.error.spaceInURL = true;
			canProcessValidation = false;
		}

		// If "canProcessValidation"
		if (canProcessValidation) {
			// Validate URL
			this.validateURL();
		} else {
			// Set this object as process completed			
			this.completed = true;
			// Analyze error and add message to 'Problems' tab
			this.addMessage();
		}
	}


	/**
	 * This method determine the URL's kind and find some potential initial error
	 */
	processURL() {
		// Check if the name of link is empty (ex.: <a href="http://domain.com"></a>)
		if (this.domTag.innerHTML.trim().length == 0) {
			this.error.nameless = true;
		}

		// Check if the A tag has no URL
		if (this.url == null || this.url.trim().length == 0) {
			this.error.emptyURL = true;
			// exit because all other tests use the URL and no URL exist
			return;
		}

		// Check if the link is an anchor in the same content (start with #)
		if (this.url.match(/^#/g) != null) {
			this.urlType.isAnchor = true;
		}

		// Check if the URL is an email
		if (this.url.match(/^\s*mailto\:/gmi) != null) {
			this.urlType.isEmail = true;
		}

		// Check if the URL is an FTP link
		if (this.url.match(/^\s*ftp\:/gmi) != null) {
			this.urlType.isFTP = true;
		}

		// Check if the URL is a local file link (file://)
		if (this.url.match(/^\s*file\:/gmi) != null) {
			this.urlType.isLocalFile = true;
		}

		if (!this.error.emptyURL) {
			// Check if the URL is excluded (from Extension's properties)
			if (vscode.workspace.getConfiguration("html-links-checker").excludedDomains.length > 0) {
				const excludedDomainsList = vscode.workspace.getConfiguration("html-links-checker").excludedDomains;
				if (excludedDomainsList.length > 0) {
					// Define default flag
					let exclError = false;
					// Keep URL in lowercase to use it in the forEach below
					const tmpURL = this.url.toLowerCase();

					// Check if all excluded domains are present in the URL
					excludedDomainsList.forEach(function (edl) {
						if (tmpURL.search(edl) > -1) {
							// Change flag value
							exclError = true;
						}
					});

					// Flag is true then this URL excluded
					if (exclError) {
						this.urlType.isExcluded = true;
					}
				}
			}
		}

		// Check if the link is relative internal (start with /)
		if (this.url.match(/^(\/{1})/g) != null) {
			this.urlType.isInternal = true;
		}

		// Check if the link is relative from the current folder
		//(start with .. or nothing and ends with web file extension: html, htm, asp, php, txt)
		// TODO: Add an extension property to define file extensions
		if (this.url.match(/(^\.(\.)?\/)|^(?!\w*\:)(\w*\.(html|htm|asp|php|txt)){1}\?*/gi) != null) {
			this.urlType.isRelative = true;
		}

		// If no other Type have been selected then the link is External or
		// internal with protocol and full path 
		if (!(this.urlType.isAnchor || this.urlType.isEmail || this.urlType.isEmpty || this.urlType.isExcluded ||
			this.urlType.isFTP || this.urlType.isInternal || this.urlType.isLocalFile || this.urlType.isRelative)) {
			this.urlType.isExternal = true;
		}
	}



	/**
	 * Main method validates each type of URL
	 */
	validateURL() {
		if (this.urlType.isAnchor) {
			this.validate_Anchor();
		}

		if (this.urlType.isEmail) {
			this.validate_Email();
		}

		if (this.urlType.isRelative) {
			this.validate_Relative();
		}

		if (this.urlType.isLocalFile) {
			this.validate_local_file();
		}

		if (this.urlType.isFTP) {
			this.validate_FTP();
		}

		if (this.urlType.isExternal) {
			// Validate External Link Accessibility
			if (vscode.workspace.getConfiguration("html-links-checker").validateExternalLinkAccessibility) {
				this.validate_External_Link_Accessibility();
			}
			this.validate_full_link();
		}

		if (this.urlType.isInternal) {
			this.validate_full_link(vscode.workspace.getConfiguration("html-links-checker").localDomain);
		}
	}

	/**
	 * Method that validates Anchor URL
	 * Check in the whole DOM (in activeTextEditor) to find the ID of the anchor
	 * If no ID or many IDs found, then error is set accordingly
	 */
	validate_Anchor() {
		const anchorLink = this.url.replace(/^#/g, '');
		const anchor = this.wholeDOM.window.document.querySelectorAll("[id='" + anchorLink + "']");
		if (anchor.length == 0) {
			this.error.noAnchor = true;
		} else if (anchor.length > 1) {
			this.error.multipleAnchors = true;
		}
		// Set this object as process completed
		this.completed = true;

		// Analyze error and add message to 'Problems' tab
		this.addMessage();
	}


	validate_External_Link_Accessibility() {
		const stringsToCheck = vscode.workspace.getConfiguration("html-links-checker").externalLinkAccessibilityStringToCheck;
		/*  TODO: This section must be improved to not validate as external link when it local
							URL using the local domain... maybe must be done before in the whole process
		let localDomain = vscode.workspace.getConfiguration("html-links-checker").localDomain;
		localDomain = localDomain.replace(/(http(s?):\/\/)?/, '');
		if (this.domTag.href.search(localDomain) == -1) { }
 		*/
		if (this.domTag.rel != 'external') {
			this.error.ExternalAccessibility = true;
		}
		let strNoPreset = true;

		for (let tmpx = 0; tmpx < stringsToCheck.length; tmpx++) {
			if (this.domTag.text.search(stringsToCheck[tmpx]) > -1) {
				strNoPreset = false;
				tmpx = stringsToCheck.length
			}
		}
		if (strNoPreset) {
			this.error.ExternalAccessibility = true;
		}
	}

	/**
	 * Method that validates Email Address
	 */
	validate_Email() {
		// Regex pattern to use to test email validation ** Get on Stackoverflow.com
		const rePattern = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
		// Remove the 'mailto:' part in the URL before test it
		const tmpURL = this.url.replace("mailto:", '');
		// Test the email address, if not OK = add error
		if (!rePattern.test(tmpURL)) {
			this.error.emailInvalid = true;
		}
		// Set this object as process completed
		this.completed = true;

		// Analyze error and add message to 'Problems' tab
		this.addMessage();
	}


	/**
	 * Method that validates relative Link
	 * TODO: Nothing is validated for the moment, only a warning is displayed to user to inform that link is relative
	 */
	validate_Relative() {
		this.warning.relativeFile = true;
		// Set this object as process completed
		this.completed = true;

		// Analyze error and add message to 'Problems' tab
		this.addMessage();
	}

	/**
	 * Method that validates local_file
	 * TODO: Nothing is validated for the moment, only a warning is displayed to user to inform that link is relative
	 */
	validate_local_file() {
		// Set this object as process completed
		this.completed = true;

		// Analyze error and add message to 'Problems' tab
		// this.addMessage();
	}

	/**
	 * Method that validates FTP 
	 * TODO: Nothing is validated for the moment, only a warning is displayed to user to inform that link is relative
	 */
	validate_FTP() {
		// Set this object as process completed
		this.completed = true;

		// Analyze error and add message to 'Problems' tab
		// this.addMessage();
	}


	/**
	 * Method that validates an external and internal link.
	 * @param {String} localDomain (optional) is the local domain to add at the beginning of the URL to create a full URL.
	 * If no value passed then the method use the URL in the object.
	 * All errors/warnings found with this URL will be set in the object
	 */
	validate_full_link(localDomain = null) {
		// Define needle's option
		const options = {
			method: vscode.workspace.getConfiguration("html-links-checker").requestMethod,
			rejectUnauthorized: false,
			followRedirect: vscode.workspace.getConfiguration("html-links-checker").validateExternalLinkFollowRedirection,
			timeout: 6000/*
			enableProxy: true,
			proxy: '',
			*/
		};
		// Copy the current object in a variable to be able to refer to it inside another object.
		const self = this;

		// Create the full URL
		// If localDomain has been provided then it's a local URL and local domain must be added
		// If not then it's external link and only use the URL in the DOM
		const fullURL = localDomain != null ? localDomain + this.url : this.url;

		// Request the URL with Promise way
		urllib.request(fullURL, options).then(function (result) {
			// Set the statusCode returned by the request
			self.statusCode = result.res.statusCode;
			// If URL have been redirected then analyze differences
			if (result.res.statusCode != 200) {
					self.error.requestError = true;
			} else if (result.res.requestUrls.length > 1) {
				//if (result.res.statusCode != 200) {
				//	self.error.requestError = true;
				//}

				// Get initial and final URL without the last / (if present)
				let initialURL = result.res.requestUrls[0].replace(/\/$/g, '');
				let finalURL = result.res.requestUrls[result.res.requestUrls.length - 1].replace(/\/$/g, '');

				// Set warning
				self.warning.redirected = true;
				// Get the new redirected URL
				self.redirectURL = finalURL;

				// Apply regex on initial and final URL for following comparaisons
				initialURL = initialURL.match(/(.*?\:\/\/)(www\.)*(.+?)$/i);
				finalURL = finalURL.match(/(.*?\:\/\/)(www\.)*(.+?)$/i);

				// Check if Protocl (HTTP(S)) has changed from the initial to final URL
				self.warning.redirectProtocolChanged = initialURL[1] == finalURL[1] ? false : true;

				// Check if WWW has been added or removed from the initial to final URL
				self.warning.redirectWWW = initialURL[2] == finalURL[2] ? false : true;
			}
			// Set this object as process completed
			self.completed = true;

			// Analyze error and add message to 'Problems' tab
			self.addMessage();

		}).catch(function (err) {
			switch (err.res.statusCode) {
				case -1:
					self.error.networkError = true;
					break;
				case -2:
					self.error.connectionTimeout = true;
					break;
				default:
					if (err.res.statusCode != 200) {
						self.error.requestError = true;
					}
					break;
			}
			// Set this object as process completed
			self.completed = true;

			// Analyze error and add message to 'Problems' tab
			self.addMessage();

		});
	}



	/**
	 * Method that checks if object contains errors and prepare the message to display to user
	 * If the object has no errors, this method won't do anything
	 * Return a text with all error for this URL
	 */
	analyzeErrors() {
		let buildMSG = '';

		if (this.error.emptyURL) {
			buildMSG += " => No URL.\n";
		}
		if (this.error.spaceInURL) {
			buildMSG += " => URL contains space(s). Remove spaces and re-execute validation.\n";
		}
		if (this.error.emailInvalid) {
			buildMSG += " => Email address is invalid.\n";
		}
		if (this.error.nameless) {
			buildMSG += " => Text link is missing.\n";
		}
		if (this.error.noAnchor) {
			buildMSG += " => the ID where link points doesn't exist.\n";
		}
		if (this.error.multipleAnchors) {
			buildMSG += " => ID where link points is not unique (more than one).\n";
		}
		if (this.error.networkError) {
			buildMSG += " => This link doesn't work/not exist.\n";
		}
		if (this.error.connectionTimeout) {
			buildMSG += " => Connection timeout.\n";
		}
		if (this.error.ExternalAccessibility) {
			buildMSG += " => Accessibility, rel=\"external\" or text is missing.\n";
		}
		if (this.error.requestError) {
			switch (this.statusCode) {
				case 401:
					buildMSG += " => Error: Authentication required, code #" + this.statusCode + ".\n";
					break;
				case 404:
					buildMSG += " => Error: Not found, code #" + this.statusCode + ".\n";
					break;
				case 407:
					buildMSG += " => Error: Proxy Authentication Required, code #" + this.statusCode + ".\n";
					break;
				case 410:
					buildMSG += " => Error: Deleted, code #" + this.statusCode + ".\n";
					break;
				case 500 - 599:
					buildMSG += " => Server Error, code #" + this.statusCode + ".\n";
					break;
				default:
					buildMSG += " => Error, code #" + this.statusCode + ".\n";
					break;
			}

		}
		if (this.warning.redirected) {
			buildMSG += " => Redirected to --> " + this.redirectURL + " <--.\n";
		}
		if (this.warning.redirectProtocolChanged) {
			//buildMSG += "	-";
		}
		if (this.warning.redirectWWW) {
			//buildMSG += "	-";
		}
		if (this.warning.relativeFile) {
			buildMSG += " => Relative URL, can't be validated.\n";
		}

		if (buildMSG == '') {
			return '';
		} else {
			buildMSG = buildMSG.replace(/\n$/gi, '');
			return 'LINK: ' + buildMSG;
		}
	}



	/**
	 * Method that adds a message in the 'Problems' tab in VSCode.
	 * @param {String} msgToShow is the message that will be displayed to user
	 */
	addMessage() {
		// Get if error or warning have been set
		const hasError = Object.keys(this.error).length > 0 ? true : false;
		const hasWarning = Object.keys(this.warning).length > 0 ? true : false;

		// If error has been set then display in 'Problems' tab
		if (hasError || hasWarning) {
			// Define the message error level
			const errorLevel = hasError ? vscode.DiagnosticSeverity.Error : vscode.DiagnosticSeverity.Warning;
			// Get the message to display
			const msgToShow = this.analyzeErrors();
			// Get the position of the element in the DOM
			const elemRange = genFunc.getDOMelementPosition(this.wholeDOM, this.domTag, true);

			// Define the Diagnostic object
			const diagElem = new vscode.Diagnostic(elemRange, msgToShow, errorLevel);
			diagElem.code = this.statusCode;
			diagElem.source = 'HTML Links Checker';
			// Add the error in the array of diagnostics and display it in the 'Problems' tab
			let myDiagnostics = Array.from(links_checker_diagColl.get(vscode.window.activeTextEditor.document.uri));
			myDiagnostics.push(diagElem);
			links_checker_diagColl.set(vscode.window.activeTextEditor.document.uri, myDiagnostics);
		} else {
			// No error to display... nothing to do for now. Maybe add a Info message.
		}
	}
	// End of the Object: validateURL
}